File: ios/Planty/Models/Actuator.swift
import Foundation

struct Actuator: Codable, Sendable, Hashable, Identifiable {
    let id: UUID
    let entityID: String
    var name: String
    let kind: ActuatorKind
    var plantIDs: [UUID]
    var policyControlEnabled: Bool = false
    var currentState: String?
    let createdAt: Date
    let updatedAt: Date
    var activeLease: ActuatorLease?
    var lightSchedule: LightSchedule?
    var fanSchedule: ActuatorSchedule?

    enum CodingKeys: String, CodingKey {
        case id
        case entityID = "entity_id"
        case name
        case kind
        case plantIDs = "plant_ids"
        case policyControlEnabled = "policy_control_enabled"
        case currentState = "current_state"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case activeLease = "active_lease"
        case lightSchedule = "light_schedule"
        case fanSchedule = "fan_schedule"
    }
}

extension Actuator {
    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        entityID = try container.decode(String.self, forKey: .entityID)
        name = try container.decode(String.self, forKey: .name)
        kind = try container.decode(ActuatorKind.self, forKey: .kind)
        plantIDs = try container.decode([UUID].self, forKey: .plantIDs)
        policyControlEnabled = try container.decodeIfPresent(Bool.self, forKey: .policyControlEnabled) ?? false
        currentState = try container.decodeIfPresent(String.self, forKey: .currentState)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        updatedAt = try container.decode(Date.self, forKey: .updatedAt)
        activeLease = try container.decodeIfPresent(ActuatorLease.self, forKey: .activeLease)
        lightSchedule = try container.decodeIfPresent(LightSchedule.self, forKey: .lightSchedule)
        fanSchedule = try container.decodeIfPresent(ActuatorSchedule.self, forKey: .fanSchedule)
    }
}

extension Actuator {
    var dailySchedule: ActuatorSchedule? {
        switch kind {
        case .fan: fanSchedule
        case .light: lightSchedule
        default: nil
        }
    }

    var isOn: Bool? {
        switch currentState?.lowercased() {
        case "on": true
        case "off": false
        default: nil
        }
    }

    var stateLabel: String {
        switch isOn {
        case true: "On"
        case false: "Off"
        case nil where currentState == "unavailable": "Unavailable"
        default: "Status unknown"
        }
    }
}

struct ActuatorSchedule: Codable, Sendable, Hashable {
    let actuatorID: UUID
    let startMinute: Int
    let endMinute: Int
    let windows: [ActuatorScheduleWindow]?
    let timezone: String
    let enabled: Bool
    let lastAppliedState: Bool?
    let lastAppliedAt: Date?
    let lastError: String?
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case actuatorID = "actuator_id"
        case startMinute = "start_minute"
        case endMinute = "end_minute"
        case windows
        case timezone
        case enabled
        case lastAppliedState = "last_applied_state"
        case lastAppliedAt = "last_applied_at"
        case lastError = "last_error"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    init(
        actuatorID: UUID,
        startMinute: Int,
        endMinute: Int,
        timezone: String,
        enabled: Bool,
        lastAppliedState: Bool?,
        lastAppliedAt: Date?,
        lastError: String?,
        createdAt: Date,
        updatedAt: Date,
        windows: [ActuatorScheduleWindow]? = nil
    ) {
        self.actuatorID = actuatorID
        self.startMinute = startMinute
        self.endMinute = endMinute
        self.windows = windows
        self.timezone = timezone
        self.enabled = enabled
        self.lastAppliedState = lastAppliedState
        self.lastAppliedAt = lastAppliedAt
        self.lastError = lastError
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    var dailyWindows: [ActuatorScheduleWindow] {
        if let windows, !windows.isEmpty { return windows }
        return [ActuatorScheduleWindow(startMinute: startMinute, endMinute: endMinute)]
    }
}

struct ActuatorScheduleWindow: Codable, Sendable, Hashable {
    let startMinute: Int
    let endMinute: Int

    enum CodingKeys: String, CodingKey {
        case startMinute = "start_minute"
        case endMinute = "end_minute"
    }
}

typealias LightSchedule = ActuatorSchedule

struct LightScheduleRequest: Codable, Sendable, Equatable {
    let startMinute: Int
    let endMinute: Int
    let windows: [ActuatorScheduleWindow]
    let timezone: String
    let enabled: Bool
    let actor: String

    enum CodingKeys: String, CodingKey {
        case startMinute = "start_minute"
        case endMinute = "end_minute"
        case windows
        case timezone, enabled, actor
    }

    init(windows: [ActuatorScheduleWindow], timezone: String, enabled: Bool, actor: String) {
        startMinute = windows.first?.startMinute ?? 0
        endMinute = windows.first?.endMinute ?? 0
        self.windows = windows
        self.timezone = timezone
        self.enabled = enabled
        self.actor = actor
    }
}

struct LightScheduleDraft: Equatable {
    var windows: [ActuatorScheduleWindowDraft]
    var timezone: String
    var enabled: Bool

    init(schedule: ActuatorSchedule?, defaultTimezone: String = TimeZone.current.identifier) {
        windows = (schedule?.dailyWindows ?? [ActuatorScheduleWindow(startMinute: 7 * 60, endMinute: 21 * 60)])
            .map(ActuatorScheduleWindowDraft.init)
        timezone = schedule?.timezone ?? defaultTimezone
        enabled = schedule?.enabled ?? true
    }

    var startMinute: Int {
        get { windows.first?.startMinute ?? 0 }
        set { windows[0].startMinute = newValue }
    }

    var endMinute: Int {
        get { windows.first?.endMinute ?? 0 }
        set { windows[0].endMinute = newValue }
    }

    var requestWindows: [ActuatorScheduleWindow] {
        windows.map { ActuatorScheduleWindow(startMinute: $0.startMinute, endMinute: $0.endMinute) }
    }

    var canSave: Bool {
        !timezone.isEmpty && windowsAreValid
    }

    mutating func addWindow() {
        guard windows.count < 12 else { return }
        let priorEnd = windows.last?.endMinute ?? 6 * 60
        let start = (priorEnd + 60) % (24 * 60)
        let end = (start + 60) % (24 * 60)
        windows.append(ActuatorScheduleWindowDraft(startMinute: start, endMinute: end))
    }

    private var windowsAreValid: Bool {
        guard !windows.isEmpty, windows.count <= 12 else { return false }
        var occupied = Array(repeating: false, count: 24 * 60)
        for window in windows {
            guard (0..<24 * 60).contains(window.startMinute),
                  (0..<24 * 60).contains(window.endMinute),
                  window.startMinute != window.endMinute else { return false }
            var minute = window.startMinute
            while minute != window.endMinute {
                guard !occupied[minute] else { return false }
                occupied[minute] = true
                minute = (minute + 1) % (24 * 60)
            }
        }
        return true
    }
}

struct ActuatorScheduleWindowDraft: Identifiable, Equatable {
    let id: UUID
    var startMinute: Int
    var endMinute: Int

    init(id: UUID = UUID(), startMinute: Int, endMinute: Int) {
        self.id = id
        self.startMinute = startMinute
        self.endMinute = endMinute
    }

    init(_ window: ActuatorScheduleWindow) {
        self.init(startMinute: window.startMinute, endMinute: window.endMinute)
    }
}

typealias ActuatorScheduleRequest = LightScheduleRequest
typealias ActuatorScheduleDraft = LightScheduleDraft

struct LightStateRequest: Codable, Sendable, Equatable {
    let isOn: Bool
    let actor: String

    enum CodingKeys: String, CodingKey {
        case isOn = "on"
        case actor
    }
}

struct LightStateResponse: Decodable, Sendable {
    let isOn: Bool

    enum CodingKeys: String, CodingKey {
        case isOn = "on"
    }
}

extension Sequence where Element == Actuator {
    func assigned(to plantID: UUID) -> [Actuator] {
        filter { $0.plantIDs.contains(plantID) }
    }
}

enum ActuatorRunDuration: Int, CaseIterable, Identifiable {
    case fiveMinutes = 300
    case tenMinutes = 600
    case fifteenMinutes = 900
    case thirtyMinutes = 1_800
    case oneHour = 3_600

    var id: Int { rawValue }

    var label: String {
        switch self {
        case .fiveMinutes: "5 minutes"
        case .tenMinutes: "10 minutes"
        case .fifteenMinutes: "15 minutes"
        case .thirtyMinutes: "30 minutes"
        case .oneHour: "1 hour"
        }
    }
}

struct ActuatorLease: Codable, Sendable, Hashable, Identifiable {
    let id: UUID
    let actuatorID: UUID
    let requestedSeconds: Int
    let deadline: Date
    let actor: String
    let source: ObservationSource
    let idempotencyKey: UUID
    let startedAt: Date?
    let stoppedAt: Date?
    let stopReason: String?
    let createdAt: Date

    var isActive: Bool { stoppedAt == nil }

    enum CodingKeys: String, CodingKey {
        case id
        case actuatorID = "actuator_id"
        case requestedSeconds = "requested_seconds"
        case deadline
        case actor
        case source
        case idempotencyKey = "idempotency_key"
        case startedAt = "started_at"
        case stoppedAt = "stopped_at"
        case stopReason = "stop_reason"
        case createdAt = "created_at"
    }
}

struct ActuatorEvent: Codable, Sendable, Hashable, Identifiable {
    let id: UUID
    let actuatorID: UUID
    let leaseID: UUID?
    let action: ActuatorEventAction
    let actor: String
    let source: ObservationSource
    let idempotencyKey: UUID?
    let detail: String?
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case actuatorID = "actuator_id"
        case leaseID = "lease_id"
        case action
        case actor
        case source
        case idempotencyKey = "idempotency_key"
        case detail
        case createdAt = "created_at"
    }
}

struct ActuatorListResponse: Codable, Sendable {
    let actuators: [Actuator]
}

struct ActuatorEventListResponse: Codable, Sendable {
    let events: [ActuatorEvent]
}

struct DiscoveredActuatorListResponse: Codable, Sendable {
    let entities: [HomeAssistantEntity]
}

struct ActuatorRegistration: Codable, Sendable, Equatable {
    let entityID: String
    let name: String
    let kind: ActuatorKind
    let plantIDs: [UUID]

    enum CodingKeys: String, CodingKey {
        case entityID = "entity_id"
        case name
        case kind
        case plantIDs = "plant_ids"
    }
}

struct ActuatorRename: Codable, Sendable {
    let name: String
    let kind: ActuatorKind
    let plantIDs: [UUID]
    let policyControlEnabled: Bool

    enum CodingKeys: String, CodingKey {
        case name
        case kind
        case plantIDs = "plant_ids"
        case policyControlEnabled = "policy_control_enabled"
    }
}

struct ActuatorStartRequest: Codable, Sendable, Equatable {
    let durationSeconds: Int
    let actor: String
    let idempotencyKey: UUID

    enum CodingKeys: String, CodingKey {
        case durationSeconds = "duration_seconds"
        case actor
        case idempotencyKey = "idempotency_key"
    }
}

struct ActuatorStopRequest: Codable, Sendable, Equatable {
    let actor: String
    let idempotencyKey: UUID

    enum CodingKeys: String, CodingKey {
        case actor
        case idempotencyKey = "idempotency_key"
    }
}

struct ActuatorStopResponse: Codable, Sendable {
    let stopped: Bool
}

File: internal/plant/actuator.go
package plant

import (
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

const (
	MaxActuatorDuration        = time.Hour
	MaxActuatorScheduleWindows = 12
)

type ActuatorKind string

const (
	ActuatorFan    ActuatorKind = "fan"
	ActuatorSwitch ActuatorKind = "switch"
	ActuatorLight  ActuatorKind = "light"
	ActuatorWater  ActuatorKind = "water"
)

type Actuator struct {
	ID                   uuid.UUID         `json:"id"`
	EntityID             string            `json:"entity_id"`
	Name                 string            `json:"name"`
	Kind                 ActuatorKind      `json:"kind"`
	PlantIDs             []uuid.UUID       `json:"plant_ids"`
	PolicyControlEnabled bool              `json:"policy_control_enabled"`
	CurrentState         string            `json:"current_state,omitempty"`
	ActiveLease          *ActuatorLease    `json:"active_lease,omitempty"`
	LightSchedule        *ActuatorSchedule `json:"light_schedule,omitempty"`
	FanSchedule          *ActuatorSchedule `json:"fan_schedule,omitempty"`
	CreatedAt            time.Time         `json:"created_at"`
	UpdatedAt            time.Time         `json:"updated_at"`
}

func (a Actuator) Valid() error {
	if strings.TrimSpace(a.Name) == "" {
		return invalid("actuator name is required")
	}
	if _, err := a.EntityDomain(); err != nil {
		return invalid("actuator entity_id must be a fan, switch, or light entity")
	}
	switch a.Kind {
	case ActuatorFan, ActuatorSwitch, ActuatorLight, ActuatorWater:
	default:
		return invalid("actuator kind must be fan, switch, light, or water")
	}
	if a.PolicyControlEnabled && a.Kind != ActuatorFan {
		return invalid("policy control is only available for fans")
	}
	if len(a.PlantIDs) == 0 {
		return invalid("actuator must be assigned to at least one plant")
	}
	seen := make(map[uuid.UUID]struct{}, len(a.PlantIDs))
	for _, id := range a.PlantIDs {
		if id == uuid.Nil {
			return invalid("actuator plant_ids must not contain an empty id")
		}
		if _, duplicate := seen[id]; duplicate {
			return invalid("actuator plant_ids must not contain duplicates")
		}
		seen[id] = struct{}{}
	}
	return nil
}

func (a Actuator) EntityDomain() (string, error) {
	domain, _, ok := strings.Cut(strings.TrimSpace(a.EntityID), ".")
	if !ok || (domain != "fan" && domain != "switch" && domain != "light") {
		return "", invalid("actuator entity_id must be a fan, switch, or light entity")
	}
	return domain, nil
}

type ActuatorSchedule struct {
	ActuatorID       uuid.UUID                `json:"actuator_id"`
	StartMinute      int                      `json:"start_minute"`
	EndMinute        int                      `json:"end_minute"`
	Windows          []ActuatorScheduleWindow `json:"windows"`
	Timezone         string                   `json:"timezone"`
	Enabled          bool                     `json:"enabled"`
	LastAppliedState *bool                    `json:"last_applied_state,omitempty"`
	LastAppliedAt    *time.Time               `json:"last_applied_at,omitempty"`
	LastError        string                   `json:"last_error,omitempty"`
	CreatedAt        time.Time                `json:"created_at"`
	UpdatedAt        time.Time                `json:"updated_at"`
}

type ActuatorScheduleWindow struct {
	StartMinute int `json:"start_minute"`
	EndMinute   int `json:"end_minute"`
}

func (s ActuatorSchedule) EffectiveWindows() []ActuatorScheduleWindow {
	if len(s.Windows) > 0 {
		return s.Windows
	}
	return []ActuatorScheduleWindow{{StartMinute: s.StartMinute, EndMinute: s.EndMinute}}
}

func (s ActuatorSchedule) Valid() error {
	if s.ActuatorID == uuid.Nil {
		return invalid("actuator_id is required")
	}
	if _, err := time.LoadLocation(strings.TrimSpace(s.Timezone)); err != nil {
		return invalid("unknown actuator schedule timezone %q", s.Timezone)
	}
	windows := s.EffectiveWindows()
	if len(windows) > MaxActuatorScheduleWindows {
		return invalid("actuator schedule may contain at most %d windows", MaxActuatorScheduleWindows)
	}
	occupied := [24 * 60]bool{}
	for _, window := range windows {
		if err := window.Valid(); err != nil {
			return err
		}
		for minute := window.StartMinute; minute != window.EndMinute; minute = (minute + 1) % (24 * 60) {
			if occupied[minute] {
				return invalid("actuator schedule windows must not overlap")
			}
			occupied[minute] = true
		}
	}
	return nil
}

func (w ActuatorScheduleWindow) Valid() error {
	if w.StartMinute < 0 || w.StartMinute > 1439 || w.EndMinute < 0 || w.EndMinute > 1439 {
		return invalid("actuator schedule minutes must be between 0 and 1439")
	}
	if w.StartMinute == w.EndMinute {
		return invalid("actuator schedule start and end must differ")
	}
	return nil
}

func (s ActuatorSchedule) WantsOn(at time.Time) (bool, error) {
	if err := s.Valid(); err != nil {
		return false, err
	}
	if !s.Enabled {
		return false, nil
	}
	location, _ := time.LoadLocation(s.Timezone)
	local := at.In(location)
	minute := local.Hour()*60 + local.Minute()
	for _, window := range s.EffectiveWindows() {
		if window.StartMinute < window.EndMinute && minute >= window.StartMinute && minute < window.EndMinute {
			return true, nil
		}
		if window.StartMinute > window.EndMinute && (minute >= window.StartMinute || minute < window.EndMinute) {
			return true, nil
		}
	}
	return false, nil
}

type LightSchedule = ActuatorSchedule

type ActuatorLease struct {
	ID               uuid.UUID  `json:"id"`
	ActuatorID       uuid.UUID  `json:"actuator_id"`
	RequestedSeconds int        `json:"requested_seconds"`
	Deadline         time.Time  `json:"deadline"`
	Actor            string     `json:"actor"`
	Source           Source     `json:"source"`
	IdempotencyKey   uuid.UUID  `json:"idempotency_key"`
	StartedAt        *time.Time `json:"started_at,omitempty"`
	StoppedAt        *time.Time `json:"stopped_at,omitempty"`
	StopReason       string     `json:"stop_reason,omitempty"`
	CreatedAt        time.Time  `json:"created_at"`
}

func (l ActuatorLease) Valid() error {
	if l.ActuatorID == uuid.Nil || l.IdempotencyKey == uuid.Nil {
		return invalid("actuator_id and idempotency_key are required")
	}
	if l.RequestedSeconds < 1 || time.Duration(l.RequestedSeconds)*time.Second > MaxActuatorDuration {
		return invalid("actuator duration must be between 1 and %d seconds", int(MaxActuatorDuration/time.Second))
	}
	if strings.TrimSpace(l.Actor) == "" {
		return invalid("actuator actor is required")
	}
	switch l.Source {
	case SourceApp, SourceAgent, SourceAutomation:
		return nil
	default:
		return fmt.Errorf("%w: unknown actuator source %q", ErrInvalid, l.Source)
	}
}

type ActuatorEvent struct {
	ID             uuid.UUID  `json:"id"`
	ActuatorID     uuid.UUID  `json:"actuator_id"`
	LeaseID        *uuid.UUID `json:"lease_id,omitempty"`
	Action         string     `json:"action"`
	Actor          string     `json:"actor"`
	Source         Source     `json:"source"`
	IdempotencyKey *uuid.UUID `json:"idempotency_key,omitempty"`
	Detail         string     `json:"detail,omitempty"`
	CreatedAt      time.Time  `json:"created_at"`
}
